core/src/main/scala/org/apache/spark/util/Utils.scala

```diff
@@ -2991,7 +2991,7 @@ private[spark] class CallerContext(
     if (CallerContext.callerContextSupported) {
       try {
         val callerContext = Utils.classForName("org.apache.hadoop.ipc.CallerContext")
-        val builder = Utils.classForName("org.apache.hadoop.ipc.CallerContext$Builder")
+        val builder: Class[_] = Utils.classForName("org.apache.hadoop.ipc.CallerContext$Builder")
```

Comment:
I don't understand how this could fix anything; it's already a `Class[_]`?

Reply:
> I don't understand how this could fix anything; it's already a `Class[_]`?

Found it return `Class[Nothing]` when debug, and Class[Nothing] cause `ClassCastException`. Use class[_], the exception doesn't happen.

Reply:
OK, that's odd, but without a generic type we should see the same thing in many places. I can't see why this one behaves differently. Maybe so, but I'm worried the cause is actually different.

Reply:
> OK, that's odd, but without a generic type we should see the same thing in many places. I can't see why this one behaves differently. Maybe so, but I'm worried the cause is actually different.

I try to read some use of `Utils.classForName`,  and I found there are about two usage scenarios: 
1. load class and not need to `getConstructor().newInstance()`
2. need to `getConstructor().newInstance()` or `getMethods`.

In second case, usually use ` :Class[_]` or write as `Utils.classForName[class]`.

Some of the uses are as follows:
https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/CallMethodViaReflection.scala#L151,L153

https://github.com/apache/spark/blob/master/core/src/test/scala/org/apache/spark/scheduler/TaskResultGetterSuite.scala#L191

https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/types/DataType.scala#L184,L185

Reply:
Hm OK that makes more sense. Ideally we'd use `Class.forName[CallerContext$Builder]` here, but I presume we can't actually directly use the class. The change is fine. You could use `Class[AnyRef]` I think to be a little more precise, but it won't matter much.

Reply:
> Hm OK that makes more sense. Ideally we'd use `Class.forName[CallerContext$Builder]` here, but I presume we can't actually directly use the class. The change is fine. You could use `Class[AnyRef]` I think to be a little more precise, but it won't matter much.

Yep, `Builder` can't be used here and `Class[AnyRef]` is better.